acme/authority_test.go

```diff
@@ -1375,8 +1383,24 @@ func TestAuthorityValidateChallenge(t *testing.T) {
 			}
 		},
 		"fail/validate-error": func(t *testing.T) test {
-			ch, err := newHTTPCh()
+			keyauth := "temp"
+			keyauthp := &keyauth
+			// Create test server that returns challenge auth
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprintf(w, "%s\r\n", *keyauthp)
+			}))
+
+			ch, err := newHTTPChWithServer(strings.TrimPrefix(ts.URL, "http://"))
+			assert.FatalError(t, err)
+
+			jwk, _, err := jose.GenerateDefaultKeyPair([]byte("pass"))
+			assert.FatalError(t, err)
+
+			thumbprint, err := jwk.Thumbprint(crypto.SHA256)
 			assert.FatalError(t, err)
+			encPrint := base64.RawURLEncoding.EncodeToString(thumbprint)
+			*keyauthp = fmt.Sprintf("%s.%s", ch.getToken(), encPrint)
+
```

Comment:
Can you just move this before httptest.NewService so you don't need to create a temporary value and all that.

Reply:
It's kind of a chicken and egg problem. httptest.NewService creates a server and sets a URL at which it can be reached. I need that URL when I create the challenge (it's stored in the challenge). However, the http Handler needs to respond with a token that comes from the challenge. So the challenge needs the server to be created, and the server needs the challenge to be created. So the way I worked around this is by using the pointer indirection. I'm all for a simpler solution if you can see one.